internal/shim-sgx/src/heap.rs

```diff
@@ -32,14 +35,25 @@ impl<const N: usize> Block<N> {
     }
 }
 
+/// The total amount of memory consumed by the ledger in bytes. Must divisible
+/// by the page size.
+///
+/// The first page can maintain 127 descriptors because it contains also the
+/// ledger's header. Subsequent pages can maintain 128 pages each. E.g. 16 pages
+/// can survive with the maximum of 2047 VMA's.
+const LEDGER_SIZE: usize = 16 * Page::SIZE;
+
+/// The amount of VMA's that the ledger has capacity to handle.
+const LEDGER_CAPACITY: usize = LEDGER_SIZE / 32 - 1;
+
 /// A heap
 pub struct Heap<'a, const N: usize>
 where
     [(); N * 64]: Sized,
 {
     blk: &'a mut Block<N>,
-    map: [u64; N * 64],
     brk: Option<NonZeroUsize>,
+    ledger: Ledger<LEDGER_CAPACITY>,
```

Comment:
Since we aren't worrying about page alignment on this type anymore, I suspect it might just be easier to pick a capacity number we think is reasonable. I think 64k is probably too big for our needs. I also think it is more reasonable to reason about in number of entries rather than by size. For example, I think 128 entries is probably more than enough for our needs once we have the ability to add pages.

This boils down to my experience with `wasmtime`. We currently have a bunch of efficiency settings turned off. However, by default `wasmtime` will attempt to map a 4GiB block with no permissions. Once we have SGX2, we don't have to allocate any pages for this. Then `wasmtime` allocates various regions within that 4GiB block (basically: rodata, data, text, stack, heap, etc). So the total number of ledger entries ends up being something like <16.

If we run out of ledger entries, we can always just fail the `mmap()` operation back to `wasmtime`. This will be observable during testing. So we should see it pretty quickly.

My vote is:

```suggestion
    ledger: Ledger<127>,
```

Reply:
I'll set it to 511.

Reply:
Fixed, and new version pushed.